temporal-sdk/src/main/java/io/temporal/internal/statemachines/WFTBuffer.java

```diff
@@ -114,21 +142,22 @@ private void flushBuffer() {
   }
 
   private void addToBuffer(HistoryEvent event) {
-    Preconditions.checkState(
-        !WFTState.None.equals(wftSequenceState),
-        "We should be inside an open WFT sequence to add to the buffer");
     wftBuffer.add(event);
   }
 
-  public List<HistoryEvent> fetch() {
+  public EventBatch fetch() {
     if (readyToFetch.size() == 1) {
       HistoryEvent event = readyToFetch.get(0);
+      Optional<HistoryEvent> wftStarted = workflowTaskCompletedEvent;
+      workflowTaskCompletedEvent = Optional.empty();
       readyToFetch.clear();
-      return Collections.singletonList(event);
+      return new EventBatch(wftStarted, Collections.singletonList(event));
     } else {
       List<HistoryEvent> result = new ArrayList<>(readyToFetch);
+      Optional<HistoryEvent> wftStarted = workflowTaskCompletedEvent;
+      workflowTaskCompletedEvent = Optional.empty();
       readyToFetch.clear();
-      return result;
+      return new EventBatch(wftStarted, result);
```

Comment:
GH won't let me suggest here because it's a whiny baby about deleted lines, but looks like this could be:

```java
    List<HistoryEvent> res;
    if (readyToFetch.size() == 1) {
      HistoryEvent event = readyToFetch.get(0);
      res = Collections.singletonList(event));
    } else {
      res = new ArrayList<>(readyToFetch);
    }
    Optional<HistoryEvent> wftStarted = workflowTaskCompletedEvent;
    workflowTaskCompletedEvent = Optional.empty();
    readyToFetch.clear();
    return new EventBatch(wftStarted, res);
```